File: 14주차/DE/110. Balanced Binary Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
 function nthNode(node){
    if(node == null){
        return 0;
    }
    
    let nthleft = nthNode(node.left);
    let nthright = nthNode(node.right);

    if(nthleft == -1 || nthright == -1 || Math.abs(nthleft - nthright) > 1) {
        return -1;
    }
    return 1 + Math.max(nthNode(node.left), nthNode(node.right))
}

function isBalanced(root) {
    if(root == null) {
        return true;
    }

    return nthNode(root) != -1;

};

File: 14주차/KM/110-balanced-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    boolean answer = true;
    public boolean isBalanced(TreeNode root) {
        checkBalanced(root);
        return answer;
    }
    public int checkBalanced(TreeNode node){
        if(node == null){
            return 0;
        }
        int left = checkBalanced(node.left);
        int right = checkBalanced(node.right);
        if(Math.abs(left - right) > 1){
            answer = false;
        }
        return Math.max(left,right)+1;
    }
}